tests/test_output_formatter.cpp

```diff
@@ -320,6 +347,22 @@ TEST_F(TableFormatter, multiple_instance_list_output)
     EXPECT_THAT(output, Eq(expected_table_output));
 }
 
+TEST_F(TableFormatter, multiple_instance_sorted_output)
+{
+    auto list_reply = construct_unsorted_list_reply();
+
+    auto expected_table_output = "Name                    State             IPv4             Image\n"
+                                 "trusty-190611-1529      Deleted           --               Ubuntu N/A\n"
+                                 "trusty-190611-1535      Stopped           --               Ubuntu N/A\n"
+                                 "trusty-190611-1539      Suspended         --               Ubuntu N/A\n"
+                                 "trusty-190611-1542      Running           --               Ubuntu N/A\n";
+
+    mp::TableFormatter table_formatter;
+    auto output = table_formatter.format(list_reply);
```

Comment:
To test treatment of petenv, you could do something like the following (did not actually try it):

```c++
EXPECT_CALL(mpt::MockSettings::mock_instance(),
        get(Eq(mp::petenv_key))).WillOnce(Return("trusty-190611-1539"));
```

And get that entry first in the expected output `trusty-190611-1539`.

Reply:
I tend to keep unit tests single-purpose, this one tests that (non-petenv) instances are sorted alphabetically. I'd like to keep it that way.

If theory we could extend the `…_petenv_first` tests a bit, but then again those test exactly what they should :)

Reply:
Unit tests are indeed supposed to test "units" of functionality, but where you decide the unit is can often subjective. IMO it would be perfectly appropriate to consider the entire sorting function as the unit here.

The lexicographical ordering is implemented elsewhere, so this really only tests whether it is being called properly on our side. If you then test the petenv comes first without lexicographical ordering of the remaining entries, nothing guarantees that they work together: you could write code that passes both tests without doing both things simultaneously ("if petenv in, put it first, else sort"). IOW the conjunction of two bits of functionality is itself functionality that requires testing. 

In practice some parsimony is needed when deciding unit boundaries, lest we end up testing single instructions :wink:

Reply:
Sure, which is why I suggested adding another test covering the conjuction of the two. And will do that.

I still find it important to know whether it's the alphabetical sorting, or the petenv precedence, or the conjunction of them failed, so all three tests make sense, IMO.

Reply:
OK, I am fine with separate testing for the conjunction. 

You may want to consider parameterized tests to reduce redundancy &ndash; they could check for both properties but be fed input in those 3 separate conditions. It would also be easy to add coverage for other limit cases too (like empty input, only petenv, only a single non-petenv).